src/lib/mongoBackend/MongoCommonUpdate.cpp

```diff
@@ -241,46 +226,35 @@ static bool equalMetadataValues(BSONObj& md1, BSONObj& md2)
 
 /* ****************************************************************************
 *
-* equalMetadataVectors -
+* equalMetadata -
 *
-* Given two vectors with the same metadata names-types, check that all the values
+* Given two metadata object assuming they have the same size (i.e. number of elements). check that all the values
 * are equal, returning false otherwise.
 *
 * Arguments are passed by reference to avoid "heavy" copies
 */
-static bool equalMetadataVectors(BSONObj& mdV1, BSONObj& mdV2)
+static bool equalMetadata(BSONObj& md1, BSONObj& md2)
 {
+  std::set<std::string>  md1Set;
 
-  bool found = false;
+  // Assuming that md1 and md2 are of equal size, we can equally use md1 or md2 for the set
+  md1.getFieldNames(md1Set);
 
-  for (BSONObj::iterator i1 = mdV1.begin(); i1.more();)
+  for (std::set<std::string>::iterator i = md1Set.begin(); i != md1Set.end(); ++i)
   {
-    BSONObj md1 = i1.next().embeddedObject();
+    std::string currentMd = *i;
 
-    for (BSONObj::iterator i2 = mdV2.begin(); i2.more();)
+    if (!md2.hasField(currentMd))
     {
-      BSONObj md2 = i2.next().embeddedObject();
-
-      /* Check metadata match */
-      if (matchMetadata(md1, md2))
-      {
-        if (!equalMetadataValues(md1, md2))
-        {
-          return false;
-        }
-        else
-        {
-          found = true;
-          break;  // loop in i2
-        }
-      }
+      return false;
     }
 
-    if (found)
+    BSONObj md1Item = getFieldF(md1, currentMd).embeddedObject();
+    BSONObj md2Item = getFieldF(md2, currentMd).embeddedObject();
+
+    if (!equalMetadataValues(md1Item, md2Item))
     {
-      /* Shortcut for early passing to the next attribute in i1 */
-      found = false;
-      continue;  // loop in i1
+      return false;
     }
   }
```

Comment:
At line 261, a little down below (you probably need to expand diff at github web to see it).